docs/developer/decision-records/2023-03-12_testing_pipeline/README.md

```diff
@@ -0,0 +1,124 @@
+    - The `EndToEndTransferCosmosDbTest` is never executed due to incorrect/missing tags
+- Conditional execution based on the existence of secrets (i.e. CosmosDB) is cumbersome (fixed in my latest PR)
+
+## Approach / Solution proposals
+
+### Avoid running untagged tests
+
+Whenever a tag inclusion is specified (`-DincludeTags="..."`), we should _only_ run tests tagged as such. We must avoid
+running untagged tests to reduce test runtime, and more clearly delineate test runs.
+For that, a small modification to the `TestConvention` in the plugins repo is necessary.
+
+### Create dedicated Gradle tasks
+
+In order to avoid having to specify the exact tag name
+as `./gradlew test -DincludeTags="PostgresqlIntegrationTest,AzureCosmosDbIntegrationTest"` we should create dedicated
+Gradle tasks to make this more robust toward refactoring. The same command could then be written as:
+
+```bash
+./gradlew clean testPostgres testAzureCosmos
+```
+
+For example, we could do the following:
+
+```java
+
+@Tag("Api")
+public class SomeApiTest {
+    @Test
+    void foo() {
+        fail("api is broken!");
+    }
+}
+```
+
+and correspondingly in the `build.gradle.kts`:
+
+```kotlin
+tasks.register<Test>("apiTest") {
+    useJUnitPlatform {
+        includeTags("Api")
+    }
+    group = "verification"
+}
+```
+
+Then, we can execute `./gradlew apiTest`, which will only run the test tagged with `@Tag("Api")`.
+
+A small Gradle Plugin creating these tasks could be implemented in the `buildSrc` directory.
+
+**We must adapt our build plugin's `TestConvention`, because currently it would exclude all `@IntegrationTest` tests
+unless the
+`includeTags` system property is specified!**
+
+### Refactoring the CI Pipeline
+
+We propose to refactor the current pipeline from its currently very linear form with a single workflow into several
+workflows:
+
+- `Verify.yaml`: all other jobs depend on this one
+    - `Checkstyle`: job that runs checkstyle
+    - `Dependency-Analysis`: runs the dependency analysis
+- `Unit-Test.yaml`: runs unit tests with coverage and uploads test reports
+- `Api-Tests.yaml`: contains a job that runs all API tests
+- `Integration-Test.yaml`: Contains the following jobs, each running the respective integration tests:
+    - `Azure-Storage`
+    - `Azure-Cosmos`: uses real Cosmos instance, runs only if `COSMOS_KEY` is present
+    - `Aws`
+    - `Daps`
+    - `Hashicorp-Vault`
+    - `Component-Test`
+- `EndToEndTest.yaml`: runs complex and potentially long running tests. may need more time.
+    - `In-Memory`: runs the `EndtoEndTransferInMemoryTest`
+    - `CosmosDb`: runs the `EndToEndTransferCosmosDbTest`, only if `COSMOS_KEY` is present
+    - `PostgreSQL`: runs the `EndToEndTransferPostgresTest`
+    - `Helm-Chart`: runs the minikube-based Helm chart test
+    - `OpenTelemetry`: runs the Opentelemetry tests
+
+Moving these jobs out into seperate workflows allows for a higher degree of parallelization, trigger configuration and
```

Comment:
I think parallelization doesn't change much, jobs are already using parallelization at the highest degree.

Reply:
Currently the `EndToEnd` job only runs the in-memory variant. Once we also run the postgres and cosmos e-2-e tests, and if we kept the current format, they would run sequentially as steps, which would blow up test runtime significantly

Reply:
Postgres E2E are running in fact, because I broke time one time or two during the last weeks, they are running within the "Postgres-Integration-Test" job but I agree that we should improve the separation

Reply:
yes they run, but not in the `End-To-End` test job, but in the `Postgres-Integration-Tests`, which is ... well... wrong :)

End-to-End tests are expected to take a long time, so parallelizing those more makes sense. Currently `End-To-End-TEsts` is a job, where all steps (PG, Az, ...) would run sequentially, but if we convert it to a workflow, PG, Az, etc could run as parallel jobs.